File: cmrdv_ws/src/cmrdv_planning/planning_codebase/midpoint/generator.py
import numpy as np
import numpy.typing as npt
import cmrdv_ws.src.cmrdv_planning.planning_codebase.raceline.raceline as raceline
from typing import List

# bounding boxes and z axes are not needed since they are constant
# if depth is given later, that will be used
# one problem that i'm worried about is if the cone count doesn't match up in pairs
class MidpointGenerator:
    def __init__(self, interpolation_number=30):
        """
        Parameters
        ----------
        interpolation_number : int
            TODO
        """

        self.PERCEP_COLOR = 2 # TODO don't hardcode PERCEP_COLOR

        self.BLUE = 1 # TODO don't hardcode constants from perceptions
        self.YELLOW = 2 # TODO don't hardcode constants from perceptions
        self.ORANGE = 3

        self.interpolation_number = interpolation_number

        self.cumulated_splines: List[raceline.Spline] = []
        self.cumulated_lengths = []

    def sorted_by_norm(self, list):
        """
        Return a copy of a list sorted by the L2 norms of its vectors in ascending order

        Parameters
        ----------
        list : npt.NDArray[npt.NDArray[np.float64]]
            list of vectors


        Returns
        -------
        npt.NDArray[npt.NDArray[np.float64]]
            a copy of the input list, sorted by norms of vectors (ascending order)
        """
        norms = np.linalg.norm(list, axis=1)
        return list[np.argsort(norms)]

    def midpoint(self, inner, outer):
        '''
        Input 2 np arrays, output their midpoints

        TODO Complete documentation
        '''

        M = 1/2 * (inner + outer)
        return M

    def generate_splines(self, midpoints):
        """
        Generate the splines fitting the given points

        Parameters
        ----------
        midpoints : npt.NDArray[npt.NDArray[np.float64]]
            list of points


        Returns
        -------
        list[raceline.Spline]
            a list of splines represented as Spline objects
        """
        # Compute splines
        [splines, cumulative_lengths] = raceline.raceline_gen(midpoints, points_per_spline=len(midpoints), loop=False) # there is no loop to do as we don't generate an entire path

        self.cumulated_splines = [*self.cumulated_splines, *splines]

        if len(self.cumulated_lengths) == 0:
            self.cumulated_lengths = cumulative_lengths
        else:
            self.cumulated_lengths = np.hstack((self.cumulated_lengths, cumulative_lengths + self.cumulated_lengths[-1]))

        return splines

    def generate_points(self, perception_data):
        """
        Generate points used for predicting a good trajectory.
        Based on the quality of perception data, some virtual midpoints might be generated.

        Parameters
        ----------
        perception_data : npt.NDArray[npt.NDArray[np.float64]]
            perception_data including cone positions and colors in the frame of the car

        Returns
        -------
        npt.NDArray[npt.NDArray[np.float64]]
            a list of points to build a trajectory from
        """
        # check that perception_data is already numpy array
        # otherwise: perception_data = np.array(perception_data)

        perception_data = np.array(perception_data)
        print(perception_data)
        # separate data into inner and outer cones
        left_indices = perception_data[:, self.PERCEP_COLOR] == self.BLUE
        right_indices = perception_data[:, self.PERCEP_COLOR] == self.YELLOW

        left_points = perception_data[left_indices, :][:, :2]
        right_points = perception_data[right_indices, :][:, :2]

        # Sort points by distance to origin
        left_points = self.sorted_by_norm(left_points)
        right_points = self.sorted_by_norm(right_points)

        #interpolate points if we only see one of a single color
        if (len(left_points) == 0 and len(right_points) > 0):
            #interpolate the other side
            for i in range(len(right_points)):
                if(i == 0):
                    #just take first cone and flip x axis
                    temp = np.copy(right_points[0])
                    temp[0] = -1*temp[0] #flip axis
                    left_points = np.append(left_points, temp)
                    left_points = np.reshape(left_points, (1, 2))
                    # print("after first",left_points)
                else: #take linear transform
                    difference = [right_points[i][0]-right_points[i-1][0], right_points[i][1]-right_points[i-1][1]]
                    difference = np.array(difference, dtype=float)
                    # print("difference",difference)
                    # print("left_points[i-1]",left_points[i-1])
                    newCone = left_points[i-1]+difference
                    # print("newCone",newCone)
                    left_points = np.vstack((left_points, newCone))
            midpoints = self.midpoint(left_points, right_points)
        elif (len(left_points) > 0 and len(right_points) == 0):
            for i in range(len(left_points)):
                if(i == 0):
                    #just take first cone and flip x axis
                    temp = np.copy(left_points[0])
                    temp[0] = -1*temp[0] #flip axis
                    right_points = np.append(right_points, temp)
                    right_points = np.reshape(right_points, (1, 2))
                else: #take linear transform
                    difference = [left_points[i][0]-left_points[i-1][0], left_points[i][1]-left_points[i-1][1]]
                    difference = np.array(difference, dtype=float)
                    newCone = right_points[i-1]+difference
                    right_points = np.vstack((right_points, newCone))
            midpoints = self.midpoint(left_points, right_points)
        else:
            # Add origin to draw path from current position
            min_len = min(np.shape(left_points)[0], np.shape(right_points)[0])
            left_points = left_points[0:min_len]
            right_points = right_points[0:min_len]
            midpoints = self.midpoint(left_points, right_points)
            midpoints = np.vstack(([0, 0], midpoints))

        return midpoints

    def interpolate_cones(self, perception_data, interpolation_number=None):
        """
        TODO Complete documentation
        """
        if interpolation_number == None:
            interpolation_number = self.interpolation_number

        spline_next = self.spline_from_cones(perception_data)

        spline: raceline.Spline = spline_next

        return spline.interpolate(interpolation_number)

    def spline_from_cones(self, perception_data):
        """
        TODO Complete documentation
        """
        print('Executing PathPlanning Midpoint Spline...')
        # define what needs to be executed at every iteration of MCL

        # Determine (virtual) midpoints
        midpoints = self.generate_points(perception_data)
        
        assert(len(midpoints) > 1)

        # We only generate one spline
        splines = self.generate_splines(midpoints)

        #assert(len(splines) == 1)
        spline = splines[0]

        return spline

File: driverless_ws/src/planning/midpoint/generator.cpp
#include "generator.hpp"


MidpointGenerator::MidpointGenerator(int interpolation_num){
    interpolation_number=interpolation_num;

}

gsl_matrix* MidpointGenerator::sorted_by_norm(gsl_matrix *list){
    gsl_matrix *res = gsl_matrix_alloc(list->size1,list->size2);
    for(int i=0;i<list->size1;i++)
        for(int j=0;j<list->size1;j++)
        gsl_matrix_set(res,i,j,gsl_matrix_get(list,i,j));

// IMPLEMENT SORT
}


gsl_matrix* midpoint(gsl_matrix *inner,gsl_matrix *outer){
    gsl_matrix *midpt = gsl_matrix_alloc(inner->size1,inner->size2);
    for(int i=0;i<midpt->size1;i++)
        for(int j=0;j<midpt->size1;j++)
        gsl_matrix_set(midpt,i,j,(gsl_matrix_get(inner,i,j)+gsl_matrix_get(outer,i,j))/2);

    return midpt;
}


std::vector<Spline> MidpointGenerator::generate_splines(gsl_matrix *midpoints){
    std::pair<std::vector<Spline>,std::vector<double>> a= raceline_gen(midpoints,std::rand(),midpoints->size2,false);
    // auto result =  raceline_gen(midpoints,std::rand(),midpoints->size2,false);
    for (auto &e:a.first){
        cumulated_splines.push_back(e);
    }
    for(auto &e:a.second){
        cumulated_lengths.push_back(e);
    }
    return a.first;
}



gsl_matrix* MidpointGenerator::generate_points(perceptionsData perceptions_data){ 
    // LEFT ==BLUE
    if (perceptions_data.bluecones.size()>0 && perceptions_data.yellowcones.size()==0){
        gsl_matrix *left = gsl_matrix_alloc(2,perceptions_data.bluecones.size());
        gsl_matrix *right = gsl_matrix_alloc(2,perceptions_data.bluecones.size());
        for(int i=0;i<perceptions_data.bluecones.size();i++){
            gsl_matrix_set(left,0,i,perceptions_data.bluecones[i].first);
            gsl_matrix_set(left,1,i,perceptions_data.bluecones[i].second);
            if(i==0){
                gsl_matrix_set(right,0,i,-1*perceptions_data.bluecones[i].first);
                gsl_matrix_set(right,1,i,perceptions_data.bluecones[i].second);
            }
            else{
                double xdiff = perceptions_data.bluecones[i].first-perceptions_data.bluecones[i-1].first;
                double ydiff = perceptions_data.bluecones[i].second-perceptions_data.bluecones[i-1].second;
                gsl_matrix_set(right,0,i,gsl_matrix_get(right,0,i-1)+xdiff);
                gsl_matrix_set(right,1,i,gsl_matrix_get(right,1,i-1)+ydiff);
            }
        }
        gsl_matrix *midpoint_mat = midpoint(left,right);
        gsl_matrix_free(left);
        gsl_matrix_free(right);
        return midpoint_mat;
    }
    if (perceptions_data.bluecones.size()==0 && perceptions_data.yellowcones.size()>0){
        gsl_matrix *left = gsl_matrix_alloc(2,perceptions_data.yellowcones.size());
        gsl_matrix *right = gsl_matrix_alloc(2,perceptions_data.yellowcones.size());
        for(int i=0;i<perceptions_data.yellowcones.size();i++){
            gsl_matrix_set(right,0,i,perceptions_data.yellowcones[i].first);
            gsl_matrix_set(right,1,i,perceptions_data.yellowcones[i].second);
            if(i==0){
                gsl_matrix_set(left,0,i,-1*perceptions_data.yellowcones[i].first);
                gsl_matrix_set(left,1,i,perceptions_data.yellowcones[i].second);
            }
            else{
                double xdiff = perceptions_data.yellowcones[i].first-perceptions_data.yellowcones[i-1].first;
                double ydiff = perceptions_data.yellowcones[i].second-perceptions_data.yellowcones[i-1].second;
                gsl_matrix_set(left,0,i,gsl_matrix_get(left,0,i-1)+xdiff);
                gsl_matrix_set(left,1,i,gsl_matrix_get(left,1,i-1)+ydiff);
            }
        }
        gsl_matrix *midpoint_mat = midpoint(left,right);
        gsl_matrix_free(left);
        gsl_matrix_free(right);
        return midpoint_mat;
    }
    double size = std::min(perceptions_data.bluecones.size(),perceptions_data.yellowcones.size());
        gsl_matrix *left = gsl_matrix_alloc(2,size+1);
        gsl_matrix *right = gsl_matrix_alloc(2,size+1);
        for(int i=0;i<perceptions_data.yellowcones.size();i++){
            gsl_matrix_set(left,0,i+1,perceptions_data.bluecones[i].first);
            gsl_matrix_set(left,1,i+1,perceptions_data.bluecones[i].second);
            gsl_matrix_set(right,0,i+1,perceptions_data.yellowcones[i].first);
            gsl_matrix_set(right,1,i+1,perceptions_data.yellowcones[i].second);

        }

        gsl_matrix *midpoint_mat = midpoint(left,right);
        gsl_matrix_free(left);
        gsl_matrix_free(right);
        return midpoint_mat;

}


std::vector<float> MidpointGenerator::interpolate_cones(gsl_matrix *perceptions_data){
    
}